### ISSUE TITLE
Encoding auto-detection doesn't auto-detect Unicode (UTF16 Little Endian), even with Byte Order Mark

### ISSUE THREAD

**Issue by AlexVallat** [dnGrep/dnGrep#998]
Using these test files:
[Encoding Test Files.zip](https://github.com/dnGrep/dnGrep/files/13448655/Encoding.Test.Files.zip)

Search for the text "Test".
Under Auto detection, UTF-8 and UTF-16 BE BOM match, but UTF-16 LE and UTF-16 LE BOM do not. I would have expected at least the LE BOM to be auto-detectable, and given that LE is the standard for UTF-16 in Windows it would be nice if that auto-detected without the BOM too.

With encoding set to Unicode it does match the two UTF-16 LE files.

**Comment by doug24** [comment]
dnGrep uses a third-party library - [UTFUnknown](https://github.com/CharsetDetector/UTF-unknown) - to detect file encoding.   There are a number of libraries, many obsolete, based on the same original source in Mozilla.  They all have many issues reported for incorrect results.  

It is a guessing game based on the characters in the file and the BOM.  Generally starting with files with a minimum of 1024 bytes works better than having very small files where there are few characters to try to identify.  It may not always guess what encoding was used to originally write the file but returns an encoding that it has the highest confidence that will read the characters that it analyzes.

dnGrep used to have its own version of this code, but it was out of date and caused many issues.  dnGrep changed to use UTFUnknown because it was more widely used and actively updated.  But as of today, it hasn't been updated in a year and half and has a number of open issues - including one yesterday for UTF16 LE not detected correctly.

Notepad++ uses a similar library (written in c++), and I like to try files there to see if is able to make a better determination.  Here is what Notepad++ was able to do:
test-utf16 BE BOM - correct
test-utf16 BE - failed
test-utf16 LE BOM - correct
test-utf16 LE - failed
test-utf8 - correct

I also tried with Notepad, Visual Studio, and VS Code, and none of them could open these UTF-16 files (BE or LE) without the BOM.  While the BOM is technically optional, it does make it difficult to determine the endian unless you know the first character is ASCII - like in xml or json.  For the real files you are trying to search - is there a reason they were written without a BOM?

In dnGrep, the problem with the test-utf16 LE BOM is something different - it is tagged as binary file.  The test for a binary file is two consecutive null bytes - which this file has.  In the past dnGrep has shipped where the binary test was four consecutive null bytes, but that was failing to detect some binary files, saying they were plain text files.  Looks like I need to come up with another test for binary files.

**Comment by AlexVallat** [comment]
The problem I actually encountered was with the UTF-16 LE BOM file. Actually a much larger file, but I reduced it to this minimal case for the issue report. I then produced the other encodings files to further explore the issue to make a full report. There was no real file without a BOM that I was trying to search, I included them for completeness.

To add to your list of examples, my editor of choice is Notepad3, and that opens all 5 files with correct auto-detection of encoding. Another interesting one is .NET CLR (e.g. `File.ReadAllText`), which auto detects both BOM files and the UTF-8 one, but neither of the ones without the BOM. Windows 10 native notepad.exe opens all but UTF-16 BE no BOM correctly.

Windows produces UTF-16 LE without BOM in some cases (for example, using /u for cmd, like `cmd /u /c dir >results.txt`), so it would probably be useful to be able to detect it.

**Comment by doug24** [comment]
`File.ReadAllText` calls a StreamReader with the flag `detectEncodingFromByteOrderMarks` set to true, and falls back to UTF-8 if this condition is not met:

```
        // Note that detectEncodingFromByteOrderMarks is a very
        // loose attempt at detecting the encoding by looking at the first
        // 3 bytes of the stream.  It will recognize UTF-8, little endian
        // unicode, and big endian unicode text, but that's it.  If neither
        // of those three match, it will use the Encoding you provided.
```

Notepad3 also uses the uchardet library - same family tree as UTF-Unknown and uchardet used in Notepad++, but must have some improvements in their version.

**Comment by AlexVallat** [comment]
Interesting! It occurs to me that you actually have an advantage over a text editor here, because while a text editor must make a single best guess as to the encoding of a file, in your case if there are multiple valid interpretations you can just search as each of them.

For example, if a file containing `일부텍스트` is saved as UTF-16 BE with no BOM, that could be `糇肽䷑꓂룒` as UTF-16 LE with no way to know which is correct short of semantically understanding that one has meaning and one is meaningless! But for your case it doesn't matter, if I'm searching for 텍스트 in the file and it matches UTF-16 BE but not LE then you have the answer. If I'm searching for "apples" then it doesn't match either and I don't even care which encoding it was.

I doubt it would be worth trying every possible codepage, but trying the various Unicode possibilities might be good. Or I guess you could have a user-configurable list of likely codepages to try? Someone in Greece might want to always try codepage 1253 in addition to Unicode, but a Russian would might want a Cyrillic code page instead. That would be quite a bit more work than just always checking possible Unicode variants, but something to consider anyway.

**Comment by doug24** [comment]
I had not thought of searching files more than once with different encodings. I hope it will not need to come to that - I've spent so much time trying to make things as efficient and fast as possible, and don't want to spend more time going through all the files multiple times.

I looked at the code in Notepad3, and while I'm not completely clear on what they are doing, I don't think the uchardet library is able to detect UTF-16 without the BOM - the library code is only looking for the BOM.  I think Notepad3 has some additional processing to handle files without a BOM.  

So, I have added some new code: after running the encoding check using UTF-Uknown, I have added an additional check if the file has no BOM, using the suggestion on the Wikipedia [Byte Order Mark](https://en.wikipedia.org/wiki/Byte_order_mark#UTF-16) page.  The idea is to look for an ACSII character with an adjacent null byte (or 3 null bytes for UTF-32).  That assumes there will be at least newline characters in the file, possibly spaces and maybe some other characters. If it is a single line file will all non-ASCII characters, there isn't much you can do other than try them all as you suggested. But it is now working for all of your test samples and for a UTF-32 sample file. 

I also discovered that the AvalonEdit code has its own encoding detection that I did not know was there. It doesn't work for Unicode without the BOM either. So, after getting the correct encoding and passing it to the control, it was getting replaced with the wrong encoding.  I made a few more changes to bypass the encoding detection in AvalonEdit.

I fixed the binary file test by going back to searching for 4 consecutive null bytes.  The binary file type I knew was failing that test turns out be another archive format, so I added it to the archives (and archive files skip the binary file test).

**Comment by AlexVallat** [comment]
Nice, thank you. I'm surprised such an algorithm isn't part of UTF-Unknown, it seems generally applicable rather than specific to dnGrep's usage. A single line of non-ASCII characters, not even spaces, would be pretty edge-case, I agree!

I take your point about performance, I had rather assumed that the limiting factor would be the reading of the file bytes off disk, rather than the searching of them. But if that's not what the actual performance characteristics look like then yeah, it would be a shame to lose a lot of performance to multiple scans when it would only be of benefit in very limited circumstances.

**Comment by doug24** [comment]
Fixed in [v4.0.189](https://github.com/dnGrep/dnGrep/releases/tag/v4.0.189.0)

**Comment by AlexVallat** [comment]
Confirmed, works for my troublesome files now, thanks!
